newsletter/2024-03-06-this-week-in-loongarch-39.md

```diff
@@ -0,0 +1,197 @@
+LoongArch 设计师觉得这还不过瘾：按照 LoongArch 过程调用约定，`$r1` 固定表示返回地址，
+因此设计师认为不可能有程序将 `csrxchg` 写掩码放入 `$r1`，于是将 `rj = 1` 的
+`csrxchg` 定义为了特例，此时 `rj` 操作数指向的 `$r1` 并不表示写掩码，也不会被读取，
+而是等价于将整个 `rd` 都写入指定的 CSR——像是 `$r1` 存放了全 1 的值一样。
+这样做了之后，仅分配一个操作码便表示了三种具体操作，从而达到了「节省编码空间」的既定目的。
+
+从个人理解和软件实现的角度，可以将 `csrrd`、`csrwr` 与 `csrxchg` 理解为操作码部分重合，但操作数格式不同的三条指令；
+也可以将其理解为同一条广义 `csrxchg` 的语法糖，此指令在 `rj = 1` 时的行为是不读 `$r1` 而以全 1 的值作为写掩码。
+在后一种理解下，通用的反汇编算法不再需要专门为这种指令实现两阶段处理，而将这部分工作留给更下游的应用——
+大部分应用不那么底层，不需要关心特权架构，也就接触不到这部分复杂度；
+而如果都到了要处理特权架构的程度了，这部分额外工作量不过是毛毛雨。
+但从龙芯官方手册描述和代码实现来看，则可以看出这些工作人员都在采用前一种理解。
+:::
+
+1 月 26 日，mengqinggang [增加了](https://sourceware.org/git/?p=binutils-gdb.git;a=commitdiff;h=969f5c0e12c18595cbd16f2df73c4630d0635dfe)
+`$s9` 为 `$fp` 寄存器的别名。早在 2021 年，LoongArch ABI 规范文档中就[已载明](https://github.com/loongson/LoongArch-Documentation/commit/de73529fa9f78e9796d97219af23c4f67b2b3de9)此别名，
+但由于最早的一批 LoongArch 上游提交内容都基于更早期的旧世界代码，而远古年代并不存在
+`$s9` 的提法——可考的最早提议是 [xen0n] 在此规范的首次公开亮相第二天[作出的](https://github.com/loongson/LoongArch-Documentation/pull/3#discussion_r684653367)，
+因此此支持长期以来都为众人所忽视了。
+
+[xen0n]: https://github.com/xen0n
+
+#### GCC {#gcc}
+
+chenxiaolong [修复了](https://gcc.gnu.org/pipermail/gcc-patches/2024-March/647255.html)一些向量测试用例的错误。
+
+Yang Yujie 为 musl libc [更换了](https://gcc.gnu.org/pipermail/gcc-patches/2024-March/647240.html)库查找路径：
+从 `/lib64` 换回了 `/lib`，禁用了 multilib 处理。
+
+[Xi Ruoyao][xry111] [重构了](https://gcc.gnu.org/pipermail/gcc-patches/2024-March/647197.html)向量浮点比较操作的测试用例，
+通过避免使用固定寄存器号的方式，使相关循环能被向量化，进而让这些测试得以正常通过了。
+
+[Xi Ruoyao][xry111] 为 gcc 也[增加了](https://gcc.gnu.org/pipermail/gcc-patches/2024-March/647193.html)将
+`fp` 称作 `s9` 的支持。
+
+#### LLVM {#llvm}
+
+LLVM 18.1.0 已于中国时间 3 月 6 日下午[正式发布](https://github.com/llvm/llvm-project/releases/tag/llvmorg-18.1.0)。请打包人们在处理此版本时注意：
+
+* 需要为 LLD 打上 [ULEB128 支持补丁](https://github.com/llvm/llvm-project/pull/83983)。
+* 需要为 LLVM 打上 [CAS 操作符号扩展补丁](https://github.com/llvm/llvm-project/pull/83750)。
+
+这些 backports 未能赶上 18.1.0，但后续的补丁版本应该会包含了，因此预计只需为
+18.1.0 这个版本做特殊处理。
+
+以下变更都对应当前开发中的分支，即 LLVM 19。
+
+[MQ-mengqing] 一个月前为 LLD [增加了](https://github.com/llvm/llvm-project/pull/81133)
+ULEB128 重定位操作的支持，这对 DWARF v5 调试信息的处理至关重要。此 PR 终于在 3 月 5 日合并了。
+
+[wangleiat] 跟进先前在第 36 期周报[报道过的](./2024-02-12-this-week-in-loongarch-36.md#gcc)
+GCC [同款修复](https://gcc.gnu.org/pipermail/gcc-patches/2024-February/645016.html)，为
+Clang 也[修复了](https://github.com/llvm/llvm-project/pull/84100)
+`__iocsrrd_h` 的返回值类型。
+
+[wangleiat] 还为 LoongArch 代码生成后端[启用了](https://github.com/llvm/llvm-project/pull/83759)
+Machine Scheduler。
+
+[SixWeining] 为解决 Firefox JavaScript 原子操作的[行为错误](https://bugzilla.mozilla.org/show_bug.cgi?id=1882301)，仿照先前
+RISC-V 的[类似修复](https://github.com/llvm/llvm-project/commit/616289ed29225c0ddfe5699c7fdf42a0fcbe0ab4)，
+为 LoongArch CAS 操作的输入[做了符号扩展](https://github.com/llvm/llvm-project/pull/83656)。
+此额外处理对拥有原生 CAS 支持的 LA664 微架构是不必要的；龙芯方面暂时没搞，[xen0n] 准备近期做掉。
+
+[leecheechen] [避免了](https://github.com/llvm/llvm-project/pull/82984)
```

Comment:
I've some doubts about this change.

IIUC an intrinsic should be simply mapped to the underlying instruction.  I.e. if an user uses the intrinsic on an invalid operand, (s)he's invoking undefined behavior and we don't need to guarantee a thing.

In GCC we don't clear those bits for `__builtin_lasx_xvpermi_q`.